### PR TITLE
fix: Makes transactions marked as DisposedBehaviour.Detach actually detach from the session pool.

### PR DESCRIPTION
**Pull request by amanda-tarafa**
The test I added was red before the fix (because of an unexpected active session) and it's now green.

I'm running the rest of the tests locally now.

### CODE REVIEW

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.ISessionPool.cs [hunk 1]
```diff
@@ -30,6 +30,7 @@ internal interface ISessionPool
             IClock Clock { get; }
             Task<PooledSession> WithFreshTransactionOrNewAsync(PooledSession session, TransactionOptions transactionOptions, CancellationToken cancellationToken);
             void Release(PooledSession session, ByteString transactionToRollback, bool deleteSession);
+            void Detach(PooledSession session);
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.ISessionPool.cs:33]: None of the implementations use the session - that sounds like it's either a bug (probably in TargetedSessionPool) or that we can remove the parameter.

**Review by jskeet — COMMENTED** [review]
I'll need to think about this more - probably opening up the code. That may well be tomorrow.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerTransaction.cs [hunk 2]
```diff
@@ -59,6 +59,8 @@ private protected SpannerTransactionBase()
     public sealed class SpannerTransaction : SpannerTransactionBase, ISpannerTransaction
     {
         private readonly List<Mutation> _mutations = new List<Mutation>();
+        // true if this transaction is used by RetriableTransaction.
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerTransaction.cs:62]: "Is used by" or "Was created by"?

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerTransaction.cs [hunk 3]
```diff
@@ -460,6 +464,15 @@ protected override void Dispose(bool disposing)
                 return;
             }
             _disposed = true;
+
+            if (_isRetriable)
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerTransaction.cs:468]: What *will* return the session back to the pool?

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.ISessionPool.cs (hunk 1 above)
Comment (jskeet) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.ISessionPool.cs:33]: Ping on this one.

Comment (amanda-tarafa) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.ISessionPool.cs:33]: It's not a bug, TargetetSessionPool does not keep track of PooledSession instances themselves, just through counters.

None of this is public so we can definetely remove the parameter. But then I think I'd like to call this method something like DetachActiveSession. Does that sound good?

**Review by amanda-tarafa — COMMENTED** [review]
Review comments addressed in second commit. Thanks!

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerTransaction.cs [hunk 4]
```diff
@@ -460,6 +464,19 @@ protected override void Dispose(bool disposing)
...
+            {
+                // If this transaction is being used by RetriableTransaction, we want to dispose of this instance
+                // but we don't want to do anything with the session, as the RetriableTransaction will attempt to
+                // reuse it with a fresh transaction.
+                // If acquiring a fresh transaction with the existing session fail, the session will be disposed
+                // and a new one with a fresh transaction will be obtained.
+                // If acquiting a fresh transaction succeeds, then the session will be diposed after the RetriableTransaction
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerTransaction.cs:475]: Nit: diposed => disposed.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerTransaction.cs (hunk 3 above)
Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerTransaction.cs:468]: Thanks, that's much clearer :)

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.ISessionPool.cs (hunk 1 above)
Comment (jskeet) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.ISessionPool.cs:33]: Either that, or we could document that the parameter is *currently* unused, but could be useful for debugging. (It feels like it's a bit of a coincidence that at the moment we don't need to say which session is being detached.)

**Review by jskeet — APPROVED** [review]
I *think* all this looks good, but I'd wait for @olavloite to approve as well.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerTransaction.cs (hunk 4 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerTransaction.cs]: Done, and also acquiting => acquiring.

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.ISessionPool.cs (hunk 1 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.ISessionPool.cs:40]: Yes, I like leaving the parameter in but documenting we don't use it. Indeed, it's totally a detail implementaiton that we don't need to use the actual session we are detaching, to detach it.

**Review by amanda-tarafa — COMMENTED** [review]
Comments address, thanks! Will wait for @olavloite review as well.

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.ISessionPool.cs [hunk 5]
```diff
@@ -30,6 +30,14 @@ internal interface ISessionPool
             IClock Clock { get; }
             Task<PooledSession> WithFreshTransactionOrNewAsync(PooledSession session, TransactionOptions transactionOptions, CancellationToken cancellationToken);
             void Release(PooledSession session, ByteString transactionToRollback, bool deleteSession);
+            /// <summary>
+            /// Detaches the given session from the pool, meaning the pool stops tracking this session as active,
+            /// but does nothing else, in particular it doesn't return the session to the pool or attempts to delete it.
```
Comment (olavloite) [inline apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.ISessionPool.cs:35]:
```suggestion
            /// but does nothing else, in particular it doesn't return the session to the pool or attempt to delete it.
```

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PooledSession.cs [hunk 6]
```diff
@@ -231,6 +231,34 @@ public void ReleaseToPool(bool forceDelete)
...
+        /// <see cref="SessionPool.CreateDetachedSession(SessionName, ByteString, ModeOneofCase)"/> (or an overload) to create an instance
+        /// of <see cref="PooledSession"/> representing the shared transaction and then call <see cref="ReleaseToPool(bool)"/> passing true
+        /// to force session deletion and clean up resources.
+        /// Else, the application can rely on Spaner service garbage collection to clean up this session once it becomes stale.
+        /// </remarks>
+        public void DetachFromPool()
+        {
+            if (MarkAsDisposed())
+            {
+                _pool.Detach(AfterReset());
```
Comment (olavloite) [inline apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PooledSession.cs:254]: Why are we calling `AfterReset()` here? Won't that remove the transaction from the session, which effectively makes it unusable from this point on? Or am I missing something here?

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerTransaction.cs [hunk 7]
```diff
@@ -460,6 +464,19 @@ protected override void Dispose(bool disposing)
                 return;
             }
             _disposed = true;
+
+            if (_isRetriable)
+            {
+                // If this transaction is being used by RetriableTransaction, we want to dispose of this instance
+                // but we don't want to do anything with the session, as the RetriableTransaction will attempt to
+                // reuse it with a fresh transaction.
+                // If acquiring a fresh transaction with the existing session fail, the session will be disposed
```
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerTransaction.cs]: Done.

**Review by amanda-tarafa — COMMENTED** [review]
@olavloite I've addressed your comments, PTAL, thanks!

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PooledSession.cs (hunk 6 above)
Comment (olavloite) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PooledSession.cs:254]: (Sorry for the slow response, I'm on-call this week).

Sorry, I thought this method was called already when you would call `Transaction.DisposeBehavior = Detach`. That is not correct, it is called when the transaction is disposed. And then it is obvious that you can't use the transaction anymore. Thanks for the explanation.

**Review by olavloite — APPROVED** [review]

**Pull request merged**